cstore_reader.c

```diff
@@ -147,78 +137,168 @@ CStoreBeginRead(const char *filename, TupleDesc tupleDescriptor,
 	readState->stripeReadContext = stripeReadContext;
 	readState->blockDataArray = blockDataArray;
 	readState->deserializedBlockIndex = -1;
+	readState->relation = relation;
 
 	return readState;
 }
 
 
 /*
- * CStoreReadFooter reads the cstore file footer from the given file. First, the
+ * CStoreReadFooter reads the cstore file footer for the given relation. First, the
  * function reads the last byte of the file as the postscript size. Then, the
  * function reads the postscript. Last, the function reads and deserializes the
  * footer.
+ *
+ * Footer data looks like following
+ * | table_footer (char[]) | postscript (char [postscriptSize]) | postscriptSize(uint8) |
+ * Parser first checks the last byte of the buffer to read postscript info that
+ * contains cstore major/minor version, cstore magic number, and byte length of
+ * the footer.
  */
 TableFooter *
-CStoreReadFooter(StringInfo tableFooterFilename)
+CStoreReadFooter(Relation relation)
 {
 	TableFooter *tableFooter = NULL;
-	FILE *tableFooterFile = NULL;
-	uint64 footerOffset = 0;
+	StringInfo footerData = NULL;
 	uint64 footerLength = 0;
 	StringInfo postscriptBuffer = NULL;
-	StringInfo postscriptSizeBuffer = NULL;
 	uint64 postscriptSizeOffset = 0;
 	uint8 postscriptSize = 0;
-	uint64 footerFileSize = 0;
 	uint64 postscriptOffset = 0;
-	StringInfo footerBuffer = NULL;
-	int freeResult = 0;
 
-	tableFooterFile = AllocateFile(tableFooterFilename->data, PG_BINARY_R);
-	if (tableFooterFile == NULL)
+	footerData = ReadFooterData(relation);
+	if (footerData == NULL)
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not open file \"%s\" for reading: %m",
-							   tableFooterFilename->data),
-						errhint("Try copying in data to the table.")));
+		return NULL;
```

Comment:
CStoreBeginWrite also calls this function now, therefore it is okay to return null. Callers are handling NULL values themselves.